lib/async_hooks.js

```diff
@@ -0,0 +1,307 @@
+
+module.exports = {
+  // Public API
+  currentId,
+  createHook,
+  // Embedder API
+  emitInit,
+  emitBefore,
+  emitAfter,
+  emitDestroy,
+  newUid,
+  setCurrentId,
+};
+
+
+// Setup the callbacks that node::AsyncWrap will call when there are hooks to
+// process. They use the same functions as the JS embedder API.
+async_wrap.setupHooks({ init, before, after, destroy });
+
+
+function createHook(fns) {
+  if (fns.init !== undefined && typeof fns.init !== 'function')
+    throw new TypeError('init must be a function');
+  if (fns.before !== undefined && typeof fns.before !== 'function')
+    throw new TypeError('before must be a function');
+  if (fns.after !== undefined && typeof fns.after !== 'function')
+    throw new TypeError('after must be a function');
+  if (fns.destroy !== undefined && typeof fns.destroy !== 'function')
+    throw new TypeError('destroy must be a function');
+
+  return new AsyncHook(fns);
+}
+
+
+function AsyncHook(fns) {
+  this[init_symbol] = fns.init;
+  this[before_symbol] = fns.before;
+  this[after_symbol] = fns.after;
+  this[destroy_symbol] = fns.destroy;
+}
+
+
+AsyncHook.prototype.enable = function enable() {
+  if (active_hooks_array.indexOf(this) !== -1)
+    return;
+  // createHook() has already enforced that the callbacks are all functions, so
+  // here simply increment the count of whether each callbacks exists or not.
+  async_hook_fields[kInit] += +!!this[init_symbol];
+  async_hook_fields[kBefore] += +!!this[before_symbol];
+  async_hook_fields[kAfter] += +!!this[after_symbol];
+  async_hook_fields[kDestroy] += +!!this[destroy_symbol];
+  async_hook_fields[kActiveHooks]++;
+  active_hooks_array.push(this);
+  return this;
+};
+
+
+AsyncHook.prototype.disable = function disable() {
+  const index = active_hooks_array.indexOf(this);
+  if (index === -1)
+    return;
+  async_hook_fields[kInit] -= +!!this[init_symbol];
+  async_hook_fields[kBefore] -= +!!this[before_symbol];
+  async_hook_fields[kAfter] -= +!!this[after_symbol];
+  async_hook_fields[kDestroy] -= +!!this[destroy_symbol];
+  async_hook_fields[kActiveHooks]--;
+  active_hooks_array.splice(index, 1);
+  return this;
+};
+
+
+// Embedder API
+
+// Usage:
+//    emitInit(number, object, string[, number[, object]])
+//
+// If parentStor is passed but parentId is undef then parentStor is ignored.
+//
+// Short circuit all checks for the common case. Which is that no hooks have
+// been set. Do this to remove performance impact for embedders (and core).
+// Even though it bypasses all the argument checks. The performance savings
+// here is critical.
+function emitInit(id, handle, type, parentId) {
+  // Return immediately if there's nothing to do. Checking for kInit covers
+  // checking for kActiveHooks. Doing this before the checks for performance.
+  if (async_hook_fields[kInit] === 0) {
+    return;
+  }
+
+  if (!Number.isSafeInteger(id) || id <= 0) {
+    throw new RangeError('id must be an unsigned integer');
```

Comment:
Given that there's not a clear concept of unsigned integers in js, perhaps this should say that it must be a non-negative integer?